# Big Data Coursework - US Treasury Bond Return Prediction

## 1. Read Data

In [25]:
import pandas as pd
import numpy as np
import pandas_datareader.data as pdr
import datetime
import yfinance as yf
import plotly.express as px

### 1.1 FRED Data

In [2]:
start = datetime.datetime(1990, 1, 1)
end = datetime.datetime(2026, 1, 31)

fred_codes = [
    'DGS10',    # 10-year Treasury yield
    'T10Y2Y',   # Term spread
    'DGS5',     # 5-Year Treasury yield
    'DGS2',     # 2-Year Treasury yield
    'TB3MS',    # 3-Month Treasury rate
    'BAA',      # Moody's BAA Corporate Bond Yield
    'AAA',      # Moody's AAA Corporate Bond Yield
    'BAMLC0A0CMEY', # ICE BofA US Corporate Index Effective Yield
    'BAMLH0A0HYM2EY', # ICE BofA US High Yield Index Effective Yield
    'TEDRATE',  # TED Spread (3M LIBOR - 3M T-Bill)
    'CPIAUCSL', # Consumer Price Index
    'T10YIE',   # 10-Year Breakeven Inflation Rate
    'UNRATE',   # Unemployment Rate
    'INDPRO',   # Industrial Production Index
    'USREC',    # NBER Recession Indicator
    'VIXCLS',   # VIX Index (Equity Volatility)
]

df_fred = pdr.DataReader(fred_codes, 'fred', start, end)

In [3]:
df_fred.head()

DGS10  T10Y2Y  DGS5  DGS2  TB3MS   BAA   AAA  BAMLC0A0CMEY  \
DATE                                                                     
1990-01-01    NaN     NaN   NaN   NaN   7.64  9.94  8.99           NaN   
1990-01-02   7.94    0.07  7.87  7.87    NaN   NaN   NaN           NaN   
1990-01-03   7.99    0.05  7.92  7.94    NaN   NaN   NaN           NaN   
1990-01-04   7.98    0.06  7.91  7.92    NaN   NaN   NaN           NaN   
1990-01-05   7.99    0.09  7.92  7.90    NaN   NaN   NaN           NaN   

            BAMLH0A0HYM2EY  TEDRATE  CPIAUCSL  T10YIE  UNRATE  INDPRO  USREC  \
DATE                                                                           
1990-01-01             NaN      NaN     127.5     NaN     5.4  61.729    0.0   
1990-01-02             NaN     0.80       NaN     NaN     NaN     NaN    NaN   
1990-01-03             NaN     0.75       NaN     NaN     NaN     NaN    NaN   
1990-01-04             NaN     0.79       NaN     NaN     NaN     NaN    NaN   
1990-01-05             NaN     0.84       NaN     NaN     NaN     NaN    NaN   

            VIXCLS  
DATE                
1990-01-01     NaN  
1990-01-02   17.24  
1990-01-03   18.19  
1990-01-04   19.22  
1990-01-05   20.11

In [4]:
# resample to monthly frequency (using end-of-month values)
df_fred_monthly = df_fred.resample('MS').last() 

df_fred_monthly.head()

DGS10  T10Y2Y  DGS5  DGS2  TB3MS    BAA   AAA  BAMLC0A0CMEY  \
DATE                                                                      
1990-01-01   8.43    0.15  8.35  8.28   7.64   9.94  8.99           NaN   
1990-02-01   8.51    0.08  8.44  8.43   7.74  10.14  9.22           NaN   
1990-03-01   8.65    0.01  8.65  8.64   7.90  10.21  9.37           NaN   
1990-04-01   9.04    0.08  9.04  8.96   7.77  10.30  9.46           NaN   
1990-05-01   8.60    0.10  8.56  8.50   7.74  10.41  9.47           NaN   

            BAMLH0A0HYM2EY  TEDRATE  CPIAUCSL  T10YIE  UNRATE   INDPRO  USREC  \
DATE                                                                            
1990-01-01             NaN     0.63     127.5     NaN     5.4  61.7290    0.0   
1990-02-01             NaN     0.61     128.0     NaN     5.3  62.2896    0.0   
1990-03-01             NaN     0.70     128.6     NaN     5.2  62.5999    0.0   
1990-04-01             NaN     0.90     128.9     NaN     5.4  62.4359    0.0   
1990-05-01             NaN     0.62     129.1     NaN     5.4  62.6258    0.0   

            VIXCLS  
DATE                
1990-01-01   25.36  
1990-02-01   21.99  
1990-03-01   19.73  
1990-04-01   19.52  
1990-05-01   17.37

In [5]:
df_fred_monthly.tail()

DGS10  T10Y2Y  DGS5  DGS2  TB3MS   BAA   AAA  BAMLC0A0CMEY  \
DATE                                                                     
2025-09-01   4.16    0.56  3.74  3.60   3.92  5.83  5.21          4.83   
2025-10-01   4.11    0.51  3.71  3.60   3.82  5.74  5.13          4.82   
2025-11-01   4.02    0.55  3.59  3.47   3.78  5.86  5.26          4.78   
2025-12-01   4.18    0.71  3.73  3.47   3.59  5.90  5.31          4.84   
2026-01-01   4.26    0.74  3.79  3.52   3.57  5.88  5.34          4.85   

            BAMLH0A0HYM2EY  TEDRATE  CPIAUCSL  T10YIE  UNRATE    INDPRO  \
DATE                                                                      
2025-09-01            6.56      NaN   324.368    2.36     4.4  101.7779   
2025-10-01            6.67      NaN       NaN    2.30     NaN  101.5163   
2025-11-01            6.56      NaN   325.031    2.23     4.5  101.9528   
2025-12-01            6.53      NaN   326.030    2.25     4.4  102.3250   
2026-01-01            6.66      NaN       NaN    2.36     NaN       NaN   

            USREC  VIXCLS  
DATE                       
2025-09-01    0.0   16.28  
2025-10-01    0.0   17.44  
2025-11-01    0.0   16.35  
2025-12-01    0.0   14.95  
2026-01-01    0.0   17.44

### 1.2 Yahoo Finance Data

In [6]:
# MOVE Index
df_move = yf.download("^MOVE", start="1990-01-01", end="2026-01-31")

/var/folders/gk/cgjc9f_s7vzgt_lt_v0jwnmh0000gn/T/ipykernel_78090/1022359267.py:2: FutureWarning: YF.download() has changed argument auto_adjust default to True
  df_move = yf.download("^MOVE", start="1990-01-01", end="2026-01-31")
[*********************100%***********************]  1 of 1 completed


In [7]:
df_move.head()

Price            Close        High         Low        Open Volume
Ticker           ^MOVE       ^MOVE       ^MOVE       ^MOVE  ^MOVE
Date                                                             
2002-11-12  120.209999  120.209999  120.209999  120.209999      0
2002-11-13  122.879997  122.879997  122.879997  122.879997      0
2002-11-14  130.289993  130.289993  130.289993  130.289993      0
2002-11-15  132.350006  132.350006  132.350006  132.350006      0
2002-11-18  129.139999  129.139999  129.139999  129.139999      0

In [8]:
# resample to monthly to match FRED data
df_move_monthly = df_move['Close'].resample('MS').last()
df_move_monthly.head()

Ticker           ^MOVE
Date                  
2002-11-01  127.519997
2002-12-01  123.889999
2003-01-01  110.889999
2003-02-01  105.360001
2003-03-01  112.089996

### 1.3 Combine Data

In [9]:
df = df_fred_monthly.merge(df_move_monthly, left_on="DATE", right_on="Date", how="outer").set_index(df_fred_monthly.index)

df.head()

DGS10  T10Y2Y  DGS5  DGS2  TB3MS    BAA   AAA  BAMLC0A0CMEY  \
DATE                                                                      
1990-01-01   8.43    0.15  8.35  8.28   7.64   9.94  8.99           NaN   
1990-02-01   8.51    0.08  8.44  8.43   7.74  10.14  9.22           NaN   
1990-03-01   8.65    0.01  8.65  8.64   7.90  10.21  9.37           NaN   
1990-04-01   9.04    0.08  9.04  8.96   7.77  10.30  9.46           NaN   
1990-05-01   8.60    0.10  8.56  8.50   7.74  10.41  9.47           NaN   

            BAMLH0A0HYM2EY  TEDRATE  CPIAUCSL  T10YIE  UNRATE   INDPRO  USREC  \
DATE                                                                            
1990-01-01             NaN     0.63     127.5     NaN     5.4  61.7290    0.0   
1990-02-01             NaN     0.61     128.0     NaN     5.3  62.2896    0.0   
1990-03-01             NaN     0.70     128.6     NaN     5.2  62.5999    0.0   
1990-04-01             NaN     0.90     128.9     NaN     5.4  62.4359    0.0   
1990-05-01             NaN     0.62     129.1     NaN     5.4  62.6258    0.0   

            VIXCLS  ^MOVE  
DATE                       
1990-01-01   25.36    NaN  
1990-02-01   21.99    NaN  
1990-03-01   19.73    NaN  
1990-04-01   19.52    NaN  
1990-05-01   17.37    NaN

In [10]:
df.tail()

DGS10  T10Y2Y  DGS5  DGS2  TB3MS   BAA   AAA  BAMLC0A0CMEY  \
DATE                                                                     
2025-09-01   4.16    0.56  3.74  3.60   3.92  5.83  5.21          4.83   
2025-10-01   4.11    0.51  3.71  3.60   3.82  5.74  5.13          4.82   
2025-11-01   4.02    0.55  3.59  3.47   3.78  5.86  5.26          4.78   
2025-12-01   4.18    0.71  3.73  3.47   3.59  5.90  5.31          4.84   
2026-01-01   4.26    0.74  3.79  3.52   3.57  5.88  5.34          4.85   

            BAMLH0A0HYM2EY  TEDRATE  CPIAUCSL  T10YIE  UNRATE    INDPRO  \
DATE                                                                      
2025-09-01            6.56      NaN   324.368    2.36     4.4  101.7779   
2025-10-01            6.67      NaN       NaN    2.30     NaN  101.5163   
2025-11-01            6.56      NaN   325.031    2.23     4.5  101.9528   
2025-12-01            6.53      NaN   326.030    2.25     4.4  102.3250   
2026-01-01            6.66      NaN       NaN    2.36     NaN       NaN   

            USREC  VIXCLS      ^MOVE  
DATE                                  
2025-09-01    0.0   16.28  77.930000  
2025-10-01    0.0   17.44  66.610001  
2025-11-01    0.0   16.35  68.949997  
2025-12-01    0.0   14.95  63.959999  
2026-01-01    0.0   17.44  59.200001

## 2. EDA & Data Preprocessing

### 2.1 Check Data Size, Type, Missing Value

In [11]:
print(df_fred_monthly.shape)
print(df_move_monthly.shape)
print(df.shape)

(433, 16)
(279, 1)
(433, 17)


In [12]:
df_fred.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 9637 entries, 1990-01-01 to 2026-01-31
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   DGS10           9025 non-null   float64
 1   T10Y2Y          9025 non-null   float64
 2   DGS5            9025 non-null   float64
 3   DGS2            9025 non-null   float64
 4   TB3MS           433 non-null    float64
 5   BAA             433 non-null    float64
 6   AAA             433 non-null    float64
 7   BAMLC0A0CMEY    7593 non-null   float64
 8   BAMLH0A0HYM2EY  7594 non-null   float64
 9   TEDRATE         7869 non-null   float64
 10  CPIAUCSL        431 non-null    float64
 11  T10YIE          5773 non-null   float64
 12  UNRATE          431 non-null    float64
 13  INDPRO          432 non-null    float64
 14  USREC           433 non-null    float64
 15  VIXCLS          9113 non-null   float64
dtypes: float64(16)
memory usage: 1.2 MB


In [13]:
df_fred_monthly.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 433 entries, 1990-01-01 to 2026-01-01
Freq: MS
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   DGS10           433 non-null    float64
 1   T10Y2Y          433 non-null    float64
 2   DGS5            433 non-null    float64
 3   DGS2            433 non-null    float64
 4   TB3MS           433 non-null    float64
 5   BAA             433 non-null    float64
 6   AAA             433 non-null    float64
 7   BAMLC0A0CMEY    350 non-null    float64
 8   BAMLH0A0HYM2EY  350 non-null    float64
 9   TEDRATE         385 non-null    float64
 10  CPIAUCSL        431 non-null    float64
 11  T10YIE          277 non-null    float64
 12  UNRATE          431 non-null    float64
 13  INDPRO          432 non-null    float64
 14  USREC           433 non-null    float64
 15  VIXCLS          433 non-null    float64
dtypes: float64(16)
memory usage: 57.5 KB


In [14]:
df_move.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 5739 entries, 2002-11-12 to 2026-01-30
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   (Close, ^MOVE)   5739 non-null   float64
 1   (High, ^MOVE)    5739 non-null   float64
 2   (Low, ^MOVE)     5739 non-null   float64
 3   (Open, ^MOVE)    5739 non-null   float64
 4   (Volume, ^MOVE)  5739 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 269.0 KB


In [15]:
df_move_monthly.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 279 entries, 2002-11-01 to 2026-01-01
Freq: MS
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ^MOVE   279 non-null    float64
dtypes: float64(1)
memory usage: 4.4 KB


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 433 entries, 1990-01-01 to 2026-01-01
Freq: MS
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   DGS10           433 non-null    float64
 1   T10Y2Y          433 non-null    float64
 2   DGS5            433 non-null    float64
 3   DGS2            433 non-null    float64
 4   TB3MS           433 non-null    float64
 5   BAA             433 non-null    float64
 6   AAA             433 non-null    float64
 7   BAMLC0A0CMEY    350 non-null    float64
 8   BAMLH0A0HYM2EY  350 non-null    float64
 9   TEDRATE         385 non-null    float64
 10  CPIAUCSL        431 non-null    float64
 11  T10YIE          277 non-null    float64
 12  UNRATE          431 non-null    float64
 13  INDPRO          432 non-null    float64
 14  USREC           433 non-null    float64
 15  VIXCLS          433 non-null    float64
 16  ^MOVE           279 non-null    float64
dtypes: floa

In [17]:
df.describe()

DGS10      T10Y2Y        DGS5        DGS2       TB3MS         BAA  \
count  433.000000  433.000000  433.000000  433.000000  433.000000  433.000000   
mean     4.233164    1.004596    3.743949    3.228568    2.714388    6.541224   
std      1.942603    0.913134    2.110034    2.267491    2.206836    1.734343   
min      0.550000   -1.060000    0.210000    0.110000    0.010000    3.160000   
25%      2.600000    0.240000    1.810000    0.940000    0.260000    5.190000   
50%      4.140000    0.910000    3.720000    3.290000    2.640000    6.320000   
75%      5.680000    1.820000    5.370000    4.970000    4.870000    7.950000   
max      9.040000    2.840000    9.040000    8.960000    7.900000   10.740000   

              AAA  BAMLC0A0CMEY  BAMLH0A0HYM2EY     TEDRATE    CPIAUCSL  \
count  433.000000    350.000000      350.000000  385.000000  431.000000   
mean     5.603557      4.790057        8.418857    0.466494  209.995032   
std      1.748652      1.598557        2.752965    0.356484   51.525537   
min      2.140000      1.790000        4.010000    0.060000  127.500000   
25%      4.070000      3.380000        6.545000    0.230000  164.550000   
50%      5.400000      4.825000        7.845000    0.380000  211.398000   
75%      7.080000      5.987500        9.395000    0.570000  242.331500   
max      9.560000      9.230000       21.810000    3.150000  326.030000   

           T10YIE      UNRATE      INDPRO       USREC      VIXCLS       ^MOVE  
count  277.000000  431.000000  432.000000  433.000000  433.000000  279.000000  
mean     2.104296    5.677262   90.331354    0.083141   19.490115   87.115627  
std      0.397182    1.741716   12.887392    0.276414    7.409734   30.774559  
min      0.110000    3.400000   60.334600    0.000000    9.510000   39.209999  
25%      1.840000    4.400000   85.953900    0.000000   13.830000   63.680000  
50%      2.220000    5.400000   94.751600    0.000000   17.470000   80.080002  
75%      2.370000    6.600000  100.267425    0.000000   23.380000  105.135002  
max      2.880000   14.800000  104.100400    1.000000   59.890000  214.000000

As seen, the column `DGS10` in  has many missing value. The reason behiind this is usually holidays (e.g., bank holiday). As later showed that we use the daily data for bond volatility calculation, we need to handle this missing value.

### 2.2 Data Cleaning and Feature Engineering

Fill the null value of fred `DGS10` using forward fill to avoid look-ahead-bias.

In [18]:
df_fred['DGS10_filled'] = df_fred['DGS10'].ffill()

Now we construct the target variable: `Monthly excess return on 10Y Treasury`.

Remember that yield != return. Yield is what we would earn if we held the bond for 10 years and interest rates never changed.

In [19]:
duration = 9  # standard approximation for a 10-Year Bond

# 1. price change (capital gain/loss due to change in yield). divide by 100 to convert to decimal
df['price_ret'] = -duration * (df['DGS10'].diff() / 100)

# 2. carry (the interest earned during the month): use the yield from the START of the month (shift 1)
df['carry_ret'] = (df['DGS10'].shift(1) / 100) / 12

# 3. excess return = total bond return minus the risk-free rate (TB3MS)
df['total_bond_ret'] = df['price_ret'] + df['carry_ret']
df['excess_return'] = df['total_bond_ret'] - ((df['TB3MS'] / 100) / 12)

The following features have missing values, so we need to handle them:

|no| variable name   |non-null count| dtype |
|--|-----------------|--------------|-------|
| 7 |  BAMLC0A0CMEY   | 350 non-null |float64|
| 8 |  BAMLH0A0HYM2EY | 350 non-null |float64|
| 9 |  TEDRATE        | 385 non-null |float64|
| 10 | CPIAUCSL       | 431 non-null |float64|
| 11 | T10YIE         | 277 non-null |float64|
| 12 | UNRATE         | 431 non-null |float64|
| 13 | INDPRO         | 432 non-null |float64|
| 16 | ^MOVE          | 279 non-null |float64|

Notes:
- `BAMLC0A0CMEY` and `BAMLH0A0HYM2EY` are entirely not used as the missing value is ~20%, and cutting all the missing value will make us lose many data --> So we will not use **Credit spread (IG)** and **Credit spread (HY)** as predictors.
- The same for `T10YIE`, that will cause us to lose almost ~40% of our data --> So we will not use **Breakeven inflation** as predictor.
- `TEDRATE` is discontinued as of January 31, 2022 --> So we will not use **TED spread** at this point. Might find an alternative of this (e.g., SOFR-based spreads), or the Default spread is already enough to measure Credit
- `^MOVE` is only available from 2002 onwards, and if we cut the row that does not have this value, we will lose almost ~40%. MOVE measures **implied volatiliy** in the U.S. Treasury bond market, based on options on Treasury futures across 2, 5, 10, and 30-year maturities to gauge market sentiment regarding interest rate risk. Therefore, on time prior to this measure is available, we can use the the **realized volatility** as an alternative (backfilling).

In [20]:
# fill minor missing macro values
df[['CPIAUCSL', 'UNRATE', 'INDPRO']] = df[['CPIAUCSL', 'UNRATE', 'INDPRO']].ffill()

# construct predictors (from project data source excel)
df['slope'] = df['T10Y2Y']
df['curvature'] = (2 * df['DGS5']) - df['DGS2'] - df['DGS10']
df['short_rate'] = df['TB3MS']
df['default_spread'] = df['BAA'] - df['AAA']
# df['credit_spread_ig] = ...
# df['credit_spread_hy] = ...
# df['ted_spread'] = ...
df['inflation_yoy'] = df['CPIAUCSL'].pct_change(periods=12)
df['unemployment'] = df['UNRATE']
df['recession_indicator'] = df['USREC']
df['industrial_prod_yoy'] = df['INDPRO'].pct_change(periods=12)
df['vix_index'] = df['VIXCLS']

# construct realized vol from DGS10 to fill the gap before 2002
# the volatility is calculated from the daily data, not the monthly data (we use higher frequency data)
# because volatility might occur in the middle of month, which cannot be captured if we only use start of month (monthly basis)
df_fred['yield_diff'] = df_fred['DGS10_filled'].diff()
# annualized vol. and rolling is by default "left-aligned"/"backward direction", so we avoid look-ahead-bias here
df_fred['realized_daily_vol_30d'] = df_fred['yield_diff'].rolling(window=30).std() * np.sqrt(252)
# use .last() so we use the volatility known at the end of the month
df['realized_bond_vol_30d'] = df_fred['realized_daily_vol_30d'].resample('MS').last()
df['bond_volatility'] = df['^MOVE'].fillna(df['realized_bond_vol_30d'])

# past 12-month bond return
df['excess_return_12m'] = df['excess_return'].rolling(window=12).sum()
df['bond_momentum'] = df['excess_return_12m'].shift(1) # lag by 1 month

Finalize the data to be used forward.

In [21]:
predictors = ['slope', 'curvature', 'short_rate', 'default_spread', 'inflation_yoy', 'unemployment', 
              'recession_indicator', 'industrial_prod_yoy', 'vix_index', 'bond_volatility', 'bond_momentum']

# lagging predictors. avoid look-ahead bias!
# we must use T-1 data to predict T returns
X = df[predictors].shift(1) # lag features by one month
y = df['excess_return']

X.info()
y.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 433 entries, 1990-01-01 to 2026-01-01
Freq: MS
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   slope                432 non-null    float64
 1   curvature            432 non-null    float64
 2   short_rate           432 non-null    float64
 3   default_spread       432 non-null    float64
 4   inflation_yoy        420 non-null    float64
 5   unemployment         432 non-null    float64
 6   recession_indicator  432 non-null    float64
 7   industrial_prod_yoy  420 non-null    float64
 8   vix_index            432 non-null    float64
 9   bond_volatility      431 non-null    float64
 10  bond_momentum        419 non-null    float64
dtypes: float64(11)
memory usage: 40.6 KB
<class 'pandas.core.series.Series'>
DatetimeIndex: 433 entries, 1990-01-01 to 2026-01-01
Freq: MS
Series name: excess_return
Non-Null Count  Dtype  
--------------  -----  
43

In [22]:
# Combine and drop the first row (NaN due to shift/pct_change) and rows with missing target
final_df = pd.concat([y, X], axis=1).dropna()

print(f"Final dataset prepared from {final_df.index.min()} to {final_df.index.max()}")
final_df.head()

Final dataset prepared from 1991-03-01 00:00:00 to 2026-01-01 00:00:00


excess_return  slope  curvature  short_rate  default_spread  \
DATE                                                                      
1991-03-01      -0.000942   0.98       0.26        5.94            1.24   
1991-04-01       0.004700   1.03       0.39        5.91            1.16   
1991-05-01      -0.001467   1.22       0.44        5.65            1.08   
1991-06-01      -0.014125   1.38       0.64        5.46            1.00   
1991-07-01       0.005817   1.34       0.66        5.57            0.95   

            inflation_yoy  unemployment  recession_indicator  \
DATE                                                           
1991-03-01       0.053125           6.6                  1.0   
1991-04-01       0.048212           6.8                  1.0   
1991-05-01       0.048099           6.7                  0.0   
1991-06-01       0.050349           6.9                  0.0   
1991-07-01       0.046959           6.9                  0.0   

            industrial_prod_yoy  vix_index  bond_volatility  bond_momentum  
DATE                                                                        
1991-03-01            -0.025780      21.23         0.610349       0.047825  
1991-04-01            -0.036187      16.88         0.735598       0.057092  
1991-05-01            -0.031106      18.24         0.685878       0.068242  
1991-06-01            -0.024950      15.93         0.626346       0.107308  
1991-07-01            -0.020371      19.55         0.574233       0.065158

In [23]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 419 entries, 1991-03-01 to 2026-01-01
Freq: MS
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   excess_return        419 non-null    float64
 1   slope                419 non-null    float64
 2   curvature            419 non-null    float64
 3   short_rate           419 non-null    float64
 4   default_spread       419 non-null    float64
 5   inflation_yoy        419 non-null    float64
 6   unemployment         419 non-null    float64
 7   recession_indicator  419 non-null    float64
 8   industrial_prod_yoy  419 non-null    float64
 9   vix_index            419 non-null    float64
 10  bond_volatility      419 non-null    float64
 11  bond_momentum        419 non-null    float64
dtypes: float64(12)
memory usage: 42.6 KB


### 2.3 Stationarity

### 2.4 Correlation 

### 2.5 PCA for Dimensionality Reduction

### 2.6 Standardization

## OLS Baseline

## Other ML Modelling